api/api/household.py

```diff
@@ -0,0 +1,61 @@
+from app import db
+from flask import abort, jsonify, make_response
+from models import Household, household_schema, households_schema
+# from . import get_timestamp
+# import uuid
+
+def read_all():
+    households = Household.query.all()
+    return households_schema.dump(households)
+
+def create(body):
+    # TODO: connexion provides me some guarantees via openAPI spec (request must have JSON body with primaryOwnerName, etc) but there are probably some other things i should check here before
+    #       committing to the database, i.e.
+    #           household = Household.query.filter(Household.primary_owner_email == body.get("primaryOwnerEmail")).first()
+    #           if household is not None:
+    #           // etc
+    #           else:
+    new_household = Household(primary_owner_name=body.get("primaryOwnerName"), primary_owner_email=body.get("primaryOwnerEmail"), secondary_owner_name=body.get("secondaryOwnerName"), secondary_owner_email=body.get("secondaryOwnerEmail"))
+    db.session.add(new_household)
+    db.session.commit()
+    return household_schema.dump(new_household)
+    
+def update(id_, household):
+    print(f"REQUEST BODY: {household}")
+    curr_household = Household.query.filter(Household.public_id == id_).one_or_none()
+
+    # TODO: try iter with setattr instead of this mess
+    # TODO: how to best check for uniqueness over column vals with uniqueness constraints & handling that
+    if curr_household is not None:
+        updated_household = household_schema.load(household, partial=True, session=db.session)
+        if updated_household.primary_owner_name:
+            curr_household.primary_owner_name = updated_household.primary_owner_name
+        if updated_household.primary_owner_email:
+            curr_household.primary_owner_email = updated_household.primary_owner_email
+        if updated_household.secondary_owner_name:
+            curr_household.secondary_owner_name = updated_household.secondary_owner_name
+        if updated_household.secondary_owner_email:
+            curr_household.secondary_owner_email = updated_household.secondary_owner_email
+        db.session.merge(curr_household)
+        db.session.commit()
+        return household_schema.dump(curr_household)
+    else:
+        abort(404, f"Household with ID {id_} not found.")
+
+def get(id_):
+    household = Household.query.filter(Household.public_id == id_).one_or_none()
+
+    if household is not None:
+        return household_schema.dump(household)
+    else:
+        abort(404, f"Could not find household with ID: {id_}")
```

Comment:
if possible, i would actually pull these methods out into their own methods and files. iirc, the pattern is technically called data access objects and data transfer objects? this would give us a standardized way of creating db rows, fetching db rows, and (hopefully) an easier way to test our endpoints. 

it's been a really long time since i've worked with python so i'm having a difficult time trying to translate go and typescript patterns into python but the general flow might look like:

```
def get(id_):
    household = Household.get(public_id)

    if household is not None:
        return household
    else:
        abort(404, f"Could not find household with ID: {id_}")
```

```
// backend/services/household/household.py or something
class Household(...):
  # init, methods, etc

def get(public_id):
  model = HouseholdModel.query.filter(...filter)
  if household is None:
    return None # or HouseholdNotFoundError
  household = new Household(...args)
  return household
```

this give us an opportunity model exactly what we want to return to the API consumer. for example, we definitely don't want to return id to the API consumer and we may not want to return createdat to the API consumer either.

Reply:
in backend/services/household/household.py

we could have methods like

`def get(public_id)`

`def create(create_household_req)` where `create_household_req` is some object/struct of only the fields we accept at household creation

`def update(update_household_req)` where `update_household_req` is some object/struct of only the fields we allow api consumer to update.

similarly, if we added some computed field to our openapi spec (idk, `primaryOwnerContactInfo`), instead of stitching it together by hand in our /get endpoint handler, we could add it to the `Household` class instead

Reply:
one things that's been hammered into my head a million times: DB schema !== API spec

Reply:
also, sorry for being terminally go/typescript-brained. i'm realizing how annoying some of these things might end up being in python

Reply:
> if possible, i would actually pull these methods out into their own methods and files. iirc, the pattern is technically called data access objects and data transfer objects? this would give us a standardized way of creating db rows, fetching db rows, and (hopefully) an easier way to test our endpoints.
> 
> it's been a really long time since i've worked with python so i'm having a difficult time trying to translate go and typescript patterns into python but the general flow might look like:
> 
> ```
> def get(id_):
>     household = Household.get(public_id)
> 
>     if household is not None:
>         return household
>     else:
>         abort(404, f"Could not find household with ID: {id_}")
> ```
> 
> ```
> // backend/services/household/household.py or something
> class Household(...):
>   # init, methods, etc
> 
> def get(public_id):
>   model = HouseholdModel.query.filter(...filter)
>   if household is None:
>     return None # or HouseholdNotFoundError
>   household = new Household(...args)
>   return household
> ```
> 
> this give us an opportunity model exactly what we want to return to the API consumer. for example, we definitely don't want to return id to the API consumer and we may not want to return createdat to the API consumer either.

i'm definitely familiar with the general style but i'll have to look and see how it can be implemented using sqlalchemy and flask because the default/style used by examples is definitely not declarative in the same way.

Reply:
> in backend/services/household/household.py
> 
> we could have methods like
> 
> `def get(public_id)`
> 
> `def create(create_household_req)` where `create_household_req` is some object/struct of only the fields we accept at household creation
> 
> `def update(update_household_req)` where `update_household_req` is some object/struct of only the fields we allow api consumer to update.
> 
> similarly, if we added some computed field to our openapi spec (idk, `primaryOwnerContactInfo`), instead of stitching it together by hand in our /get endpoint handler, we could add it to the `Household` class instead

as of now, i use the openAPI spec to declare all fields and then, for a given path and request body definition, define the fields actually required so this is happening already, though semi-implicitly and in a way that i honestly don't really care for. defining explicit data models for business logic vs internal representation is definitely something i've thought about, i just didn't know where to put that. in retrospect, i assume that shoving the full entity representations into the openAPI spec doesn't make sense when that's supposed to be the publicly communicated/represented form of the data.

Reply:
> i assume that shoving the full entity representations into the openAPI spec doesn't make sense when that's supposed to be the publicly communicated/represented form of the data.

yeah, that's what i mean by DB schema !== API spec. i can probs put together a small typescript demo if that would help?

Reply:
i'm thinking about the data flow as something like:
database <=> sqlalchemy/marshmallow models <=> data access objects/data  <=> api consumer/client

where sqlalchemy and marshmallow help you with reads/writes from the db and the data access objects help keep a consistent API for ur API responses.

Reply:
okay okay okay i've been told "type hints + dataclasses" help make python sane

https://realpython.com/python-data-classes/

Reply:
i spent some of the morning looking over this problem and it looks like just implementing the access patterns over the sqlalchemy models directly might be the way to go afaict. sqlalchemy ORM really mixes everything together which cuts against an explicitly separate data access implementation/logic.

Reply:
recent commits added my attempt of a CRUD mixin class that abstracts the operations you described and I implemented them over the household api endpoints to match the pseudo code you outlined. The only real difference right now is that I converted to exceptions instead of checking existence because I figured that might allow for refined error messages at the API level based on exception type.

I also removed non-public properties from the serialization implementation.